[IMPROVED] Better attempt at delivering messages to queue subscriptions

kozlovic opened the pull request:
This PR is based out of #633. It imroves parsing QRSID so that the
TestRouteQueueSemantics test now passes (when dealing with malformed
QRSID).
A test similar to what is reported in #632 was also added. This
test however, uncovers a race condition that will be fixed in a
separate PR.

Resolves #632

derekcollison commented on server/client.go:
```diff
@@ -1092,53 +1097,69 @@ func (c *client) processMsg(msg []byte) {
 			}
-			return
+			isRouteQsub = true
+			// for queue subscription try hard to deliver a message at least once.
```
We want to maintain strict at-most-once semantics. I do not think this violates it, but we should update the comment IMO.

kozlovic replied:
I guess you commented from original PR, which in this PR the comments were already changed. Anyway, this comment has been removed.

derekcollison commented on server/client.go:
```diff
@@ -1092,53 +1097,69 @@ func (c *client) processMsg(msg []byte) {
+			// we didn't make a delivery attempt, because either a subscriber limit
+			// was exceeded or a subscription is already gone.
+			// So, let a code below find yet another matching subscription.
+			// We are at risk that a message might go forth and back
```
back and forth

kozlovic replied:
Fixed

derekcollison commented on server/client.go:
```diff
@@ -1092,53 +1097,69 @@ func (c *client) processMsg(msg []byte) {
+			// So, let a code below find yet another matching subscription.
+			// We are at risk that a message might go forth and back
+			// between routes during these attempts, but at the end
+			// it shall either be delivered (at most once) or drop.
```
dropped

kozlovic replied:
Fixed

derekcollison commented on server/client.go:
```diff
@@ -1092,53 +1097,69 @@ func (c *client) processMsg(msg []byte) {
+			// We are at risk that a message might go forth and back
+			// between routes during these attempts, but at the end
+			// it shall either be delivered (at most once) or drop.
+			c.Debugf("Re-sending message of a detached queue sid %s", c.pa.sid)
```
This causes concern here, I don't think we are re-sending a message here correct?

kozlovic replied:
Again, comment had been previously changed (and this debug statement removed).

derekcollison commented on server/client.go:
```diff
@@ -1147,11 +1168,22 @@ func (c *client) processMsg(msg []byte) {
-				c.deliverMsg(sub, mh, msg)
+			// Iterate over all subscribed clients starting at a random index
+			// until we find one that's able to deliver a message.
+			// Drop a message on the floor if there are noone.
```
none

kozlovic replied:
Fixed.

kozlovic commented:
@derekcollison I believe I have addressed the issues regarding comments. We need to think if we want this behavior to be merged or stick with the old one. I believe that this change still honor the at-most and there should not be duplicates. But original behavior was "correct" too (since we don't guarantee at-least-once, but at-most-once).

derekcollison commented:
As long as we can guarantee that we are still at most-once then LGTM. If we can not be sure of that then we need to revisit the changes.

kozlovic commented:
@derekcollison After careful review and to the best of my knowledge, as the code stands in this PR, we still honor at-most-once delivery but do a better job at delivery messages to a queue group. If you approve, please approve the PR (it is currently in Request Changes status).
